### ISSUE TITLE
torch.cdist returns inconsistent result

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#37734]
```python
import torch

A, B = map(torch.load('bug.pt').get, ['A', 'B'])
print(A.shape, B.shape) # torch.Size([95, 39]) torch.Size([295, 39])

D = torch.cdist(A, B)
print(D[75, 233]) # tensor(0., grad_fn=<SelectBackward>)

D_ = torch.cdist(A[75].unsqueeze(0), B[233].unsqueeze(0)).squeeze()
print(D_) # tensor(0.0004, grad_fn=<SqueezeBackward0>)
```
Both ways should have same precision properties. It is very strange that they return different results: 0.0004 is quite a large difference

[bug.pt.gz](https://github.com/pytorch/pytorch/files/4570875/bug.pt.gz)

**Comment by vadimkantorov** [comment]
The first way also produces `nan` gradient, since `cdist` happens to produce `norm=0`, whereas it's not even exactly zero (as seen in the second way)

**Comment by vadimkantorov** [comment]
I wonder if the cause is different block reduction configuration used, and then the problem is really nasty

**Comment by ngimel** [comment]
The issue of `nan` gradient if cdist is 0 is being addressed in #37337. As for the numeric differences, for the full computation a matrix-based approach is used, which is much faster but also more susceptible to numerical issues. For the second case, a custom kernel is used. So the difference in results is expected. https://github.com/pytorch/pytorch/blob/090ea775c91cd94c847e3a1b62ca0e6a3a533a01/aten/src/ATen/native/Distance.cpp#L89-L96

**Comment by vadimkantorov** [comment]
Isn’t it too much difference for numerical impression?

**Comment by vadimkantorov** [comment]
using `double` bring difference to 1e-12 range. Maybe worth noting the precision properties somehow in the docs? I wonder if someone has calculated precision properties of both methods

**Comment by albanD** [comment]
Note that running this same code on my mac gives:
```
albandes-mbp:test albandes$ py foo.py
torch.Size([95, 39]) torch.Size([295, 39])
tensor(0.0003, grad_fn=<SelectBackward>)
tensor(0.0004, grad_fn=<SqueezeBackward0>)
```

Which is not great but not as bad :D 

But I think it is still worth adding a note in the doc stating that the mm version might be less precise.

**Comment by ngimel** [comment]
It's also 0.0003 on the GPU, so mkl must be doing something funny.

**Comment by albanD** [comment]
We have seen as well that the number of used threads for CPU can have a large impact on the precision of these functions in the worst case scenarios.

**Comment by vadimkantorov** [comment]
In reality, precision can be even worse. Probably it's exactly zero because it's negative and then clamped.

**Comment by vadimkantorov** [comment]
> We have seen as well that the number of used threads for CPU can have a large impact on the precision of these functions in the worst case scenarios.

It's only 39 reduced values here per mm cell. I hope only one thread/cell is used :)

It can also be that different width of the accumulator is used?

**Comment by vadimkantorov** [comment]
About error analysis: I guess it'd be sth similar to https://www.johndcook.com/blog/2008/09/28/theoretical-explanation-for-numerical-results/

**Comment by albanD** [comment]
> It can also be that different width of the accumulator is used?

About this, @ngimel do we have documented anywhere the single vs double precision accumulators that are used for CPU/GPU operators?

**Comment by ngimel** [comment]
No (in part because we are not too consistent about it, in part because for calls offloaded to libraries, such as matrix multiplication, it would depend on the underlying library)

**Comment by vadimkantorov** [comment]
A similar issue in sklearn also with fingers pointed at MKL: https://github.com/scikit-learn/scikit-learn/issues/11711

**Comment by vadimkantorov** [comment]
I guess actionable "feature requests" are:
1) Determine if MKL has worse precision than CUDA matrix multiply
2) Document these quirks
3) Does keops have better precision?

**Comment by vadimkantorov** [comment]
Oh wow. I didn't know cdist supported computation algorithm choosing: https://github.com/pytorch/pytorch/issues/42479
